File: leetcode/maximal_square.py
# Given a 2D binary matrix filled with 0's and 1's, find the largest square 
# containing all 1's and return its area.
# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    # @param {character[][]} matrix
    # @return {integer}
    def maximalSquare(self, matrix):
        result = 0
        for i in xrange(len(matrix)):
            for j in xrange(len(matrix[i])):
                matrix[i][j] = int(matrix[i][j])
                if i and j and matrix[i][j] == 1:
                    matrix[i][j] = 1 + min(matrix[i-1][j-1], matrix[i][j-1],  \
                                           matrix[i-1][j])
                result = max(result, matrix[i][j])
        return result * result

if __name__ == "__main__":
    solution = Solution()
    print solution.maximalSquare([list("1111"), list("1111"), list("1111")])

File: leetcode/maximal_square.js
/* Given a 2D binary matrix filled with 0's and 1's, find the largest square 
 * containing all 1's and return its area.
 * Time Complexity: O(mn)
 * Space Complexity: O(1)
 */

/**
 * @param {character[][]} matrix
 * @return {number}
 */
var maximalSquare = function(matrix) {
    for (var result = 0, i = 0; i < matrix.length; ++i) {
        for (var j = 0; j < matrix[i].length; ++j) {
            matrix[i][j] = parseInt(matrix[i][j]);
            if (i > 0 && j > 0 && matrix[i][j] === 1)
                matrix[i][j] = 1 + Math.min(matrix[i-1][j-1], matrix[i][j-1], 
                                            matrix[i-1][j]);
            result = Math.max(result, matrix[i][j]);
        }
    }
    return result * result;
};
